cmd/tailscale/main.go

```diff
@@ -451,26 +473,71 @@ func (a *App) runBackend(ctx context.Context) error {
 			case OAuth2Event:
 				go b.backend.Login(e.Token)
 			case BeExitNodeEvent:
-				state.Prefs.SetAdvertiseExitNode(bool(e))
-				go b.backend.SetPrefs(state.Prefs)
+				go func() {
+					a.mu.Lock()
+					prefs, err := a.editPrefs(ctx, state, func(currPrefs ipn.PrefsView, updatedPrefs *ipn.MaskedPrefs) {
+						routes := currPrefs.AdvertiseRoutes()
+						routes.AppendTo([]netip.Prefix{
+							netip.MustParsePrefix("0.0.0.0/0"),
+							netip.MustParsePrefix("::/0"),
+						})
+						updatedPrefs.AdvertiseRoutes = routes.AsSlice()
+						updatedPrefs.SetAdvertiseExitNode(bool(e))
+						updatedPrefs.AdvertiseRoutesSet = true
+					})
+					prefsView := prefs.View()
+					state.Prefs = &prefsView
+					a.mu.Unlock()
+					if prefs.AdvertisesExitNode() != bool(e) {
+						log.Printf("localapi failed to edit prefs for BeExitNodeEvent")
+					}
```

Comment:
if a function returns (T, error), conventionally (unless documented to the contrary) T is not valid for inspection when err is non-nil. So this could crash.